### DISCUSSION TITLE
Ability to set localStorage.debug programmatically / automatically to help troubleshoot CI hangs?

### DISCUSSION THREAD

**Discussion by xeger** [cypress-io/cypress#24658] [Questions and Help]
I have a _very_ high rate of Cypress suites that hang in CI -- as much as 5%. The problem happens in multiple test suites And using `DEBUG` in the parent process has not yielded any information.

I've verified that if I set `localStorage.debug` in the driver process [as documented here](https://docs.cypress.io/api/events/catalog-of-events#Logging-All-Events), the console debug logs are printed in the parent process via the `cypress:server:browsers:electron console message` channel. They're not pretty, but they exist! This lets me see the driver's debug trace in my CI output (via the parent process's debug prints to stderr).

Problem: the Cypress documentation claims that to enable `localStorage.debug` I need to type the command into my developer console, then reload the page. Running in CI with a headless browser, I can't do this! What is the equivalent step -- if any! -- that would let me enable driver debug output programmatically, e.g. at driver startup time?

I'm hoping I could hook into a `Cypress.on` event, but I'm afraid I would need to hook a very low-level event ("on browser started but driver not yet loaded?") to accomplish what I want. Seems like Cypress could offer a feature here -- a way to let me auto-enable driver debug traces for emergency situations like this.

**Comment by portyu9** [comment]
Hi @xeger 

There still isn't a documented Cypress API that programmatically sets the Cypress driver's own `localStorage.debug` before the driver starts in a headless `cypress run`.

The two debugging mechanisms run in different contexts:

- `DEBUG=cypress:*` enables Cypress Node/backend debug namespaces.
- `localStorage.debug = 'cypress:*'` enables browser-side Cypress driver logging and is documented as a DevTools workflow.

Trying to set it with `onBeforeLoad` would not be equivalent: that operates on the application's window, not early enough on Cypress's own runner context to enable driver initialization logging.

For a CI-only hang, I'd therefore use targeted backend namespaces rather than `cypress:*` if the output is too large, and collect the CI machine's CPU/memory information as well.

On current Cypress, if the run is recorded to Cypress Cloud, Test Replay/Cloud diagnostics are also much more useful for browser-side console/network information than they were when this question was posted in 2022.

Please mark this answer as accepted if it helped, thank you.
